### PR TITLE
add mint test for metadata size restrictions

### PR DESCRIPTION
**Pull request by aead**
This change adds a test for the metadata size restriction of AWS.
AWS allows user-defined metadata of 2KB and a total HTTP header
size of 8KB.

Fixes minio/mint#119

### CODE REVIEW

**Comment by poornas** [comment]
@aead, you will need to use refactored PutObject call because #811 was merged.

**Comment by aead** [comment]
Thanks @poornas

File: functional_tests.go [hunk 1]
```diff
@@ -262,6 +262,59 @@ func testMakeBucketError() {
 	successLogger(function, args, startTime).Info()
 }
 
+func testMetadataSizeLimit() {
+	startTime := time.Now()
+	function := "PutObjectWithMetadata(bucketName, objectName, reader, metadata)"
```
Comment (poornas) [inline functional_tests.go]: the signature of the function needs to change.

File: functional_tests.go [hunk 2]
```diff
@@ -262,6 +262,59 @@ func testMakeBucketError() {
...
+
+	const HeaderSizeLimit = 8 * 1024
+	const UserMetadataLimit = 2 * 1024
+
+	metadata := make(map[string]string)
+	metadata["X-Amz-Meta-Mint-Test"] = string(bytes.Repeat([]byte("m"), 1+UserMetadataLimit-len("X-Amz-Meta-Mint-Test")))
+	args["metadata"] = fmt.Sprint(metadata)
+	_, err = c.PutObject(bucketName, objectName, bytes.NewReader(nil), 0, &minio.PutObjectOptions{UserMetadata: metadata})
+	if err == nil {
+		failureLog(function, args, startTime, "", "Created object with too much user-defined metadata successfully", nil).Fatal()
```
Comment (poornas) [inline functional_tests.go]: Perhaps the message could be worded as
Created object with user-defined metadata exceeding metadata size limits.

File: functional_tests.go [hunk 3]
```diff
@@ -262,6 +262,59 @@ func testMakeBucketError() {
...
+	}
+
+	metadata = make(map[string]string)
+	metadata["X-Amz-Mint-Test"] = string(bytes.Repeat([]byte("m"), 1+HeaderSizeLimit-len("X-Amz-Mint-Test")))
+	args["metadata"] = fmt.Sprint(metadata)
+	_, err = c.PutObject(bucketName, objectName, bytes.NewReader(nil), 0, &minio.PutObjectOptions{UserMetadata: metadata})
+	if err == nil {
+		failureLog(function, args, startTime, "", "Created object with too much metadata successfully", nil).Fatal()
```
Comment (poornas) [inline functional_tests.go]: similar suggestion for "Created object with too much metadata successfully" as previous comment.

File: functional_tests.go [hunk 4]
```diff
@@ -262,6 +262,59 @@ func testMakeBucketError() {
...
+	err = c.MakeBucket(bucketName, "us-east-1")
+	if err != nil {
+		failureLog(function, args, startTime, "", "Make bucket failed", err).Fatal()
+	}
+
+	const HeaderSizeLimit = 8 * 1024
+	const UserMetadataLimit = 2 * 1024
+
+	metadata := make(map[string]string)
+	metadata["X-Amz-Meta-Mint-Test"] = string(bytes.Repeat([]byte("m"), 1+UserMetadataLimit-len("X-Amz-Meta-Mint-Test")))
```
Comment (ebozduman) [inline functional_tests.go]: A comment here would be good to explain that this is a negative test and hence a metadata will be created with `1 + <allowed_metadata_limit_bytes> - len("X-Amz-Meta-Mint-Test")`.

File: functional_tests.go (hunk 2 above)
Comment (ebozduman) [inline functional_tests.go]: `too much` = > `too many` if the message wording is going to have some kind of a quantification.

File: functional_tests.go [hunk 5]
```diff
@@ -262,6 +262,62 @@ func testMakeBucketError() {
...
+	}
+
+	// Meta-data (headers) greater than the 8 KB limit of AWS - PUT calls with this meta-data should fail
+	metadata = make(map[string]string)
+	metadata["X-Amz-Mint-Test"] = string(bytes.Repeat([]byte("m"), 1+HeaderSizeLimit-len("X-Amz-Mint-Test")))
+	args["metadata"] = fmt.Sprint(metadata)
+	_, err = c.PutObject(bucketName, objectName, bytes.NewReader(nil), 0, &minio.PutObjectOptions{UserMetadata: metadata})
+	if err == nil {
+		failureLog(function, args, startTime, "", "Created object with headers exceeding header size limits", nil).Fatal()
+	}
```
Comment (poornas) [inline functional_tests.go]: Could you pls call successLogger() as well when the test passes? Otherwise looks good.

Comment (aead) [reply functional_tests.go]: Sure, thanks.

**Comment by aead** [comment]
Sorry, Github's 'update branch' does a merge commit not a rebase -.-

**Pull request merged**